bftengine/src/bftengine/messages/CheckpointMsg.cpp

```diff
@@ -11,31 +11,48 @@
 
 #include "CheckpointMsg.hpp"
 #include "assertUtils.hpp"
+#include "Crypto.hpp"
+#include "SigManager.hpp"
 
 namespace bftEngine {
 namespace impl {
 
-CheckpointMsg::CheckpointMsg(ReplicaId senderId,
+CheckpointMsg::CheckpointMsg(ReplicaId genReplica,
                              SeqNum seqNum,
                              const Digest& stateDigest,
                              bool stateIsStable,
                              const concordUtils::SpanContext& spanContext)
-    : MessageBase(senderId, MsgCode::Checkpoint, spanContext.data().size(), sizeof(Header)) {
+    : MessageBase(genReplica,
+                  MsgCode::Checkpoint,
+                  spanContext.data().size(),
+                  sizeof(Header) + SigManager::instance()->getMySigLength()) {
   b()->seqNum = seqNum;
   b()->stateDigest = stateDigest;
   b()->flags = 0;
+  b()->genReplicaId = genReplica;
   if (stateIsStable) b()->flags |= 0x1;
   std::memcpy(body() + sizeof(Header), spanContext.data().data(), spanContext.data().size());
 }
 
+void CheckpointMsg::sign() {
+  auto sigManager = SigManager::instance();
+  sigManager->sign(body(), sizeof(Header), body() + sizeof(Header) + spanContextSize(), sigManager->getMySigLength());
+}
+
 void CheckpointMsg::validate(const ReplicasInfo& repInfo) const {
   ConcordAssert(type() == MsgCode::Checkpoint);
   ConcordAssert(senderId() != repInfo.myId());
 
-  if (size() < sizeof(Header) + spanContextSize() || (!repInfo.isIdOfReplica(senderId())) ||
+  auto sigManager = SigManager::instance();
+  auto sigLen = sigManager->getMySigLength();
+
+  if (size() < sizeof(Header) + spanContextSize() + sigLen || (!repInfo.isIdOfReplica(senderId())) ||
       (seqNumber() % checkpointWindowSize != 0) || (digestOfState().isZero()))
-    throw std::runtime_error(__PRETTY_FUNCTION__);
+    throw std::runtime_error(__PRETTY_FUNCTION__ + std::string(": basic validations"));
 
+  if (!sigManager->verifySig(
+          idOfGeneratedReplica(), body(), sizeof(Header), body() + sizeof(Header) + spanContextSize(), sigLen))
+    throw std::runtime_error(__PRETTY_FUNCTION__ + std::string(": verifySig"));
```

Comment:
does this mean the signature is also verified in the RO ?

Reply:
There was a problem with the validation in ReadOnlyReplica.hpp in messageHandler. The same was present in ReplicaForStateTransfer.hpp's messageHandler. Those are all the message handlers that did not get the fix from [PR 916](https://github.com/vmware/concord-bft/pull/916).